File: src/mwrap-ast.cc
/*
 * mwrap-ast.cc
 *   Recursive routines for AST printing, identifier construction,
 *   and destruction.
 *
 * Copyright (c) 2007  David Bindel
 * See the file COPYING for copying permissions
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <cassert>
#include "mwrap-ast.h"


/* -- Add inheritance relationship -- */
/*
 * We represent inheritance relationships as a list of child classes
 * associated with each parent.  This makes sense when we generate the
 * type casting routines, since we only allow casts to the parent
 * types.  But it's the opposite of how inheritance relations are
 * specified syntactically.
 */

map<string, InheritsDecl*> class_decls;


void add_inherits(const char* childname, InheritsDecl* ilist)
{
    for (; ilist; ilist = ilist->next) {
        InheritsDecl*& children = class_decls[ilist->name];
        children = new InheritsDecl(mwrap_strdup(childname), children);
    }
}


/* -- Add scalar type data -- */
/*
 * The names of the numeric scalar types are stored in the set scalar_decls.
 * We provide a basic list, but the user can specify more -- the only real
 * constraint is there has to be a meaningful typecast to/from a double.
 */

set<string> scalar_decls;
set<string> cscalar_decls;
set<string> zscalar_decls;
set<string> mxarray_decls;


void init_scalar_types()
{
    const char* scalar_types[] = {
        "double", "float", 
        "long", "int", "short", "char", 
        "ulong", "uint", "ushort", "uchar",
	"int32_t", "int64_t", "uint32_t", "uint64_t",
        "bool", "size_t", "ptrdiff_t", NULL};

    for (const char** s = scalar_types; *s; ++s)
        add_scalar_type(*s);
}


char *promote_int(char* name)
{
  /* Detect C99 types: int32_t, int64_t, uint32_t, uint64_t */
  if( strcmp(name,"int32_t") == 0 ) mw_use_int32_t = 1;
  if( strcmp(name,"int64_t") == 0 ) mw_use_int64_t = 1;
  if( strcmp(name,"uint32_t") == 0 ) mw_use_uint32_t = 1;
  if( strcmp(name,"uint64_t") == 0 ) mw_use_uint64_t = 1;

  if( strcmp(name,"ulong") == 0 ) mw_use_ulong = 1;
  if( strcmp(name,"uint") == 0 ) mw_use_uint = 1;
  if( strcmp(name,"ushort") == 0 ) mw_use_ushort = 1;
  if( strcmp(name,"uchar") == 0 ) mw_use_uchar = 1;

  /* Promote integers */
  if( mw_promote_int == 1 )
    {      
      if( strcmp(name,"uint") == 0 ) mw_use_ulong = 1;
      if( strcmp(name,"int") == 0 ) return strdup("long");
      if( strcmp(name,"uint") == 0 ) return strdup("ulong");
    }
  if( mw_promote_int == 2 )
    {      
      if( strcmp(name,"uint") == 0 ) mw_use_ulong = 1;
      if( strcmp(name,"ulong") == 0 ) mw_use_ulong = 1;
      if( strcmp(name,"int") == 0 ) return strdup("long");
      if( strcmp(name,"long") == 0 ) return strdup("long");
      if( strcmp(name,"uint") == 0 ) return strdup("ulong");
      if( strcmp(name,"ulong") == 0 ) return strdup("ulong");
    }
  if( mw_promote_int == 3 )
    {      
      if( strcmp(name,"int") == 0 ) mw_use_int32_t = 1;
      if( strcmp(name,"long") == 0 ) mw_use_int64_t = 1;
      if( strcmp(name,"uint") == 0 ) mw_use_uint32_t = 1;
      if( strcmp(name,"ulong") == 0 ) mw_use_uint64_t = 1;
      if( strcmp(name,"int") == 0 ) return strdup("int32_t");
      if( strcmp(name,"long") == 0 ) return strdup("int32_t");
      if( strcmp(name,"uint") == 0 ) return strdup("uint64_t");
      if( strcmp(name,"ulong") == 0 ) return strdup("uint64_t");
    }
  if( mw_promote_int == 4 )
    {      
      if( strcmp(name,"int") == 0 ) mw_use_int64_t = 1;
      if( strcmp(name,"long") == 0 ) mw_use_int64_t = 1;
      if( strcmp(name,"uint") == 0 ) mw_use_uint64_t = 1;
      if( strcmp(name,"ulong") == 0 ) mw_use_uint64_t = 1;
      if( strcmp(name,"int") == 0 ) return strdup("int64_t");
      if( strcmp(name,"long") == 0 ) return strdup("int64_t");
      if( strcmp(name,"uint") == 0 ) return strdup("uint64_t");
      if( strcmp(name,"ulong") == 0 ) return strdup("uint64_t");
    }
  return name;
}



void add_scalar_type(const char* name)
{
    scalar_decls.insert(name);
}


void add_cscalar_type(const char* name)
{
    cscalar_decls.insert(name);
}


void add_zscalar_type(const char* name)
{
    zscalar_decls.insert(name);
}


void add_mxarray_type(const char* name)
{
    mxarray_decls.insert(name);
}


bool is_scalar_type(const char* name)
{
    return (scalar_decls.find(name) != scalar_decls.end());
}


bool is_cscalar_type(const char* name)
{
    return (cscalar_decls.find(name) != cscalar_decls.end());
}


bool is_zscalar_type(const char* name)
{
    return (zscalar_decls.find(name) != zscalar_decls.end());
}


bool is_mxarray_type(const char* name)
{
    return (mxarray_decls.find(name) != mxarray_decls.end());
}


/* -- Print the AST -- */
/*
 * Print a human-readable translation of the abstract syntax tree to
 * the output.  This is only used for generating comprehensible comments
 * in the C code.
 */


void print(FILE* fp, Expr* e)
{
    if (!e)
        return;
    fprintf(fp, "%s", e->value);
    if (e->next) {
        fprintf(fp, ", ");
        print(fp, e->next);
    }
}


void print(FILE* fp, TypeQual* q)
{
    if (!q)
        return;
    if (q->qual == 'a') {
        fprintf(fp, "[");
        print(fp, q->args);
        fprintf(fp, "]");
    } else
        fprintf(fp, "%c", q->qual);
}

void print_devicespec(FILE* fp, Var* v)
{
    if (v->devicespec == 'g')
        fprintf(fp, "gpu ");
}

void print_iospec(FILE* fp, Var* v)
{
    if (v->iospec == 'o')
        fprintf(fp, "output ");
    else if (v->iospec == 'b')
        fprintf(fp, "inout ");
}


void print_var(FILE* fp, Var* v)
{
    fprintf(fp, "%s", v->basetype);
    print(fp, v->qual);
    fprintf(fp, " %s", v->name);
}


void print_args(FILE* fp, Var* v)
{
    if (!v)
        return;
    print_devicespec(fp, v);
    print_iospec(fp, v);
    print_var(fp, v);
    if (v->next) {
        fprintf(fp, ", ");
        print_args(fp, v->next);
    }
}


void print(FILE* fp, Func* f)
{
    if (!f)
        return;
    if (f->ret) {
        print_var(fp, f->ret);
        fprintf(fp, " = ");
    }
    if (f->thisv)
        fprintf(fp, "%s->%s.", f->thisv, f->classv);
    fprintf(fp, "%s(", f->funcv);
    print_args(fp, f->args);
    fprintf(fp, ");\n");
}


/* -- ID string -- */
/*
 * The ID string is a compressed text representation of the AST for
 * a function.  We replace all the variable names by the letter 'x'
 * so that when we have the same function call with different arguments,
 * we will generate the same signature.
 */


string id_string(Expr* e)
{
    if (!e)
        return "";
    return "x" + id_string(e->next);
}


string id_string(TypeQual* q)
{
    if (!q)
        return "";
    if (q->qual == 'a')
        return "[" + id_string(q->args) + "]";
    else
        return (string() + q->qual);
}


string id_string(Var* v)
{
    if (!v)
        return "";

    string name;
    if (v->devicespec == 'c')
        name += "c ";
    else if (v->devicespec == 'g')
        name += "g ";

    if (v->iospec == 'i')
        name += "i ";
    else if (v->iospec == 'o')
        name += "o ";
    else
        name += "io ";
    name += promote_int(v->basetype);
    name += id_string(v->qual);
    if (v->tinfo == VT_const) {
        name += " ";
        name += v->name;
    }
    if (v->next) {
        name += ", ";
        name += id_string(v->next);
    }
    return name;
}


string id_string(Func* f)
{
    if (!f)
        return "";

    string name;
    if (f->ret) {
        name += id_string(f->ret);
        name += " = ";
    }
    if (f->thisv) {
        name += f->thisv;
        name += "->";
        name += f->classv;
        name += ".";
    }
    name += f->funcv;
    name += "(";
    name += id_string(f->args);
    name += ")";
    return name;
}


/* -- Destroy an AST -- */


void destroy(Expr* e)
{
    if (!e)
        return;
    destroy(e->next);
    delete[] e->value;
    delete e;
}


void destroy(TypeQual* q)
{
    if (!q)
        return;
    destroy(q->args);
    delete q;
}


void destroy(Var* v)
{
    if (!v) 
        return;
    destroy(v->next);
    destroy(v->qual);
    delete[] v->basetype;
    delete[] v->name;
    delete v;
}


void destroy(Func* f)
{
    while (f) {
        Func* oldf = f;
        f = f->next;

        destroy(oldf->same_next);
        destroy(oldf->ret);
        if (oldf->thisv)
            delete[] oldf->thisv;
        if (oldf->classv)
            delete[] oldf->classv;
        delete[] oldf->funcv;
        destroy(oldf->args);
        delete oldf;
    }
}


void destroy(InheritsDecl* i)
{
    if (!i)
        return;
    destroy(i->next);
    delete i;
}


void destroy_inherits()
{
    map<string,InheritsDecl*>::iterator i = class_decls.begin();
    for (; i != class_decls.end(); ++i) {
        destroy(i->second);
        i->second = NULL;
    }
}

File: python/mwrap_ast.py
"""
mwrap_ast.py — AST nodes, type registries, and helpers for mwrap.

Copyright (c) 2007-2008  David Bindel
See the file COPYING for copying permissions

Converted to Python by Zydrunas Gimbutas (2026),
with assistance from Claude Code / Claude Opus 4.6 (Anthropic).
"""

from enum import IntEnum
from dataclasses import dataclass, field
from typing import Optional


# ---------------------------------------------------------------------------
# VT_* classification codes
# ---------------------------------------------------------------------------

class VT(IntEnum):
    unk       = 0
    obj       = 1
    array     = 2
    carray    = 3
    zarray    = 4
    rarray    = 5
    scalar    = 6
    cscalar   = 7
    zscalar   = 8
    string    = 9
    mx        = 10
    p_obj     = 11
    p_scalar  = 12
    p_cscalar = 13
    p_zscalar = 14
    r_obj     = 15
    r_scalar  = 16
    r_cscalar = 17
    r_zscalar = 18
    const     = 19


# ---------------------------------------------------------------------------
# iospec / VT classification helpers (single source of truth)
# ---------------------------------------------------------------------------

def iospec_is_input(c):   return c in ('i', 'b')
def iospec_is_output(c):  return c in ('o', 'b')
def iospec_is_inonly(c):  return c == 'i'

def is_array(tinfo):
    return tinfo in (VT.array, VT.carray, VT.zarray)

def is_obj(tinfo):
    return tinfo in (VT.obj, VT.p_obj, VT.r_obj)

def complex_tinfo(v):
    return v.tinfo in (VT.carray, VT.zarray,
                       VT.cscalar, VT.zscalar,
                       VT.r_cscalar, VT.r_zscalar,
                       VT.p_cscalar, VT.p_zscalar)

def nullable_return(f):
    return (f.ret and f.ret[0].tinfo in (
        VT.string, VT.array, VT.carray, VT.zarray,
        VT.p_scalar, VT.p_cscalar, VT.p_zscalar))


# ---------------------------------------------------------------------------
# AST node dataclasses
# ---------------------------------------------------------------------------

@dataclass
class Expr:
    value: str
    input_label: int = -1


@dataclass
class TypeQual:
    qual: str          # '*', '&', 'a' (array), 'r' (array ref)
    args: list = field(default_factory=list)  # list[Expr]


@dataclass
class Var:
    devicespec: str    # 'c' (cpu) or 'g' (gpu)
    iospec: str        # 'i','o','b'
    basetype: str
    qual: Optional[TypeQual]
    name: str
    tinfo: int = VT.unk
    input_label: int = -1
    output_label: int = -1


@dataclass
class Func:
    thisv: Optional[str]
    classv: Optional[str]
    funcv: str
    fname: str
    line: int
    fort: bool = False
    id: int = -1
    args: list = field(default_factory=list)  # list[Var]
    ret: list = field(default_factory=list)   # list[Var] (0 or 1 elements)
    same: list = field(default_factory=list)  # list[Func] — duplicate signatures


# ---------------------------------------------------------------------------
# MwrapContext — all mutable state in one object
# ---------------------------------------------------------------------------

class MwrapContext:
    def __init__(self):
        # CLI flags
        self.mw_use_gpu = False
        self.mw_generate_catch = False
        self.mw_use_c99_complex = False
        self.mw_use_cpp_complex = False
        self.mw_promote_int = 0

        # Type registries
        self.scalar_decls = set()
        self.cscalar_decls = set()
        self.zscalar_decls = set()
        self.mxarray_decls = set()
        self.class_decls = {}   # str -> list[str]

        # Type usage tracking
        self.mw_use_int32_t = 0
        self.mw_use_int64_t = 0
        self.mw_use_uint32_t = 0
        self.mw_use_uint64_t = 0
        self.mw_use_longlong = 0
        self.mw_use_ulonglong = 0
        self.mw_use_ulong = 0
        self.mw_use_uint = 0
        self.mw_use_ushort = 0
        self.mw_use_uchar = 0

    def init_scalar_types(self):
        self.scalar_decls.clear()
        for t in ("double", "float",
                  "long", "int", "short", "char",
                  "ulong", "uint", "ushort", "uchar",
                  "int32_t", "int64_t", "uint32_t", "uint64_t",
                  "bool", "size_t", "ptrdiff_t"):
            self.scalar_decls.add(t)

    def is_scalar_type(self, name):    return name in self.scalar_decls
    def is_cscalar_type(self, name):   return name in self.cscalar_decls
    def is_zscalar_type(self, name):   return name in self.zscalar_decls
    def is_mxarray_type(self, name):   return name in self.mxarray_decls

    def add_scalar_type(self, name):   self.scalar_decls.add(name)
    def add_cscalar_type(self, name):  self.cscalar_decls.add(name)
    def add_zscalar_type(self, name):  self.zscalar_decls.add(name)
    def add_mxarray_type(self, name):  self.mxarray_decls.add(name)


# ---------------------------------------------------------------------------
# Functions that need context
# ---------------------------------------------------------------------------

def promote_int(ctx, name: str) -> str:
    if name == "int32_t":  ctx.mw_use_int32_t = 1
    if name == "int64_t":  ctx.mw_use_int64_t = 1
    if name == "uint32_t": ctx.mw_use_uint32_t = 1
    if name == "uint64_t": ctx.mw_use_uint64_t = 1
    if name == "ulong":    ctx.mw_use_ulong = 1
    if name == "uint":     ctx.mw_use_uint = 1
    if name == "ushort":   ctx.mw_use_ushort = 1
    if name == "uchar":    ctx.mw_use_uchar = 1

    if ctx.mw_promote_int == 1:
        if name == "uint":  ctx.mw_use_ulong = 1
        if name == "int":   return "long"
        if name == "uint":  return "ulong"
    elif ctx.mw_promote_int == 2:
        if name == "uint":  ctx.mw_use_ulong = 1
        if name == "ulong": ctx.mw_use_ulong = 1
        if name == "int":   return "long"
        if name == "long":  return "long"
        if name == "uint":  return "ulong"
        if name == "ulong": return "ulong"
    elif ctx.mw_promote_int == 3:
        if name == "int":   ctx.mw_use_int32_t = 1
        if name == "long":  ctx.mw_use_int64_t = 1
        if name == "uint":  ctx.mw_use_uint32_t = 1
        if name == "ulong": ctx.mw_use_uint64_t = 1
        if name == "int":   return "int32_t"
        if name == "long":  return "int32_t"
        if name == "uint":  return "uint64_t"
        if name == "ulong": return "uint64_t"
    elif ctx.mw_promote_int == 4:
        if name == "int":   ctx.mw_use_int64_t = 1
        if name == "long":  ctx.mw_use_int64_t = 1
        if name == "uint":  ctx.mw_use_uint64_t = 1
        if name == "ulong": ctx.mw_use_uint64_t = 1
        if name == "int":   return "int64_t"
        if name == "long":  return "int64_t"
        if name == "uint":  return "uint64_t"
        if name == "ulong": return "uint64_t"
    return name


def add_inherits(ctx, childname: str, parents: list):
    """Register childname as child of each parent in parents list."""
    for parent in parents:
        if parent not in ctx.class_decls:
            ctx.class_decls[parent] = []
        ctx.class_decls[parent].insert(0, childname)


# ---------------------------------------------------------------------------
# ID string (canonical signature for deduplication)
# ---------------------------------------------------------------------------

def _id_expr(args: list) -> str:
    return "x" * len(args)


def _id_qual(q: Optional[TypeQual]) -> str:
    if not q:
        return ""
    if q.qual == 'a':
        return "[" + _id_expr(q.args) + "]"
    return q.qual


def _id_var_single(ctx, v: Var) -> str:
    name = ""
    if v.devicespec == 'c':
        name += "c "
    elif v.devicespec == 'g':
        name += "g "

    io = v.iospec
    if io == 'i':   name += "i "
    elif io == 'o': name += "o "
    else:           name += "io "

    name += promote_int(ctx, v.basetype)
    name += _id_qual(v.qual)
    if v.tinfo == VT.const:
        name += " " + v.name
    return name


def _id_var(ctx, vars: list) -> str:
    return ", ".join(_id_var_single(ctx, v) for v in vars)


def id_string(ctx, f) -> str:
    if not f:
        return ""
    name = ""
    if f.ret:
        name += _id_var(ctx, f.ret) + " = "
    if f.thisv:
        name += f.thisv + "->" + f.classv + "."
    name += f.funcv + "(" + _id_var(ctx, f.args) + ")"
    return name


# ---------------------------------------------------------------------------
# AST printing (for C comments)
# ---------------------------------------------------------------------------

def _print_expr(args: list) -> str:
    return ", ".join(e.value for e in args)


def _print_qual(q):
    if not q:
        return ""
    if q.qual == 'a':
        return "[" + _print_expr(q.args) + "]"
    return q.qual


def _print_devicespec(v):
    if v.devicespec == 'g':
        return "gpu "
    return ""


def _print_iospec(v):
    m = {'o': "output ", 'b': "inout "}
    return m.get(v.iospec, "")


def _print_var(v):
    return v.basetype + _print_qual(v.qual) + " " + v.name


def _print_args(vars: list) -> str:
    return ", ".join(
        _print_devicespec(v) + _print_iospec(v) + _print_var(v)
        for v in vars
    )


def print_func(f):
    """Human-readable translation of Func AST (for C comments)."""
    if not f:
        return ""
    s = ""
    if f.ret:
        s += _print_var(f.ret[0]) + " = "
    if f.thisv:
        s += f"{f.thisv}->{f.classv}."
    s += f"{f.funcv}({_print_args(f.args)});\n"
    return s
